### ISSUE TITLE
Ark + CRD + Myself caused an outage

### ISSUE THREAD

**Issue by gianrubio** [heptio/ark#539]
## TL;DR

Last week I had 2 outages caused by : ark + crd + myself. I haven't found a similar issue in the kubernetes communitty so I'm gonna post it here. I suspect it's related to CRD storing a large amount of data in etcd.

## My misconfiguration

3 weeks ago I configured ark using the [default values](https://github.com/heptio/ark/blob/master/examples/aws/00-ark-config.yaml). This default values execute disk snapshots every 30 minutes x number of pvc you have in your cluster. After 20 days running ark I noticed that the number of snapshots on my aws account reached 9k volumes snapshots. 

When I realised these unnecessary snapshots I changed my schedule to run only once a day, after that ark deleted all the snapshots from aws but keep the records on `deletebackuprequests.ark.heptio.com` crd.

## The outage

Last Tuesday I had the first outage where all the etcd and master nodes, that run in the same instances, suddenly increased the cpu and memory consumption. I haven't had enough time to investigate the issue and it was fixed increasing the master nodes from m4.large to m4.xlarge.

So it happened again last Friday, the same symptoms, however I had time to investigate the issue. Took almost 5 hours to find the root cause of my downtime.  The apiserver and etcd kept crashing, using all the cpu and restarting every 5 minutes in all the masters.E ven scaling the master to 1 replica didn't reduce the high load in the master node.

At that time I logged into the master node up and read the logs, that showed to me that ark `CRD` was crashing k8s api (or the api was crashing by itself).  Here's the first log

```
E0608 09:47:56.797099       7 runtime.go:66] Observed a panic: &errors.errorString{s:"kill connection/stream"} (kill connection/stream)
/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:230
/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:114
/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/requestinfo.go:45
/usr/local/go/src/net/http/server.go:1942
/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/request/requestcontext.go:110
/usr/local/go/src/net/http/server.go:1942
/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/wrap.go:41
/usr/local/go/src/net/http/server.go:1942
/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/handler.go:198
/usr/local/go/src/net/http/server.go:2568
/usr/local/go/src/net/http/server.go:3088
<autogenerated>:312
/usr/local/go/src/net/http/h2_bundle.go:4331
/usr/local/go/src/net/http/h2_bundle.go:4611
/usr/local/go/src/runtime/asm_amd64.s:2197
E0608 09:47:56.817509       7 wrap.go:34] APIServer panic'd on GET /apis/ark.heptio.com/v1/deletebackuprequests?resourceVersion=0: kill connection/stream
```

I figured out that ark+crd were overloading api and etcd so my first action was to scale down ark. Even so I had the same cluster behaviour of my cluster, except that I got other error from the apiserver. 

``` I0608 10:26:04.966850       7 wrap.go:42] GET /apis/ark.heptio.com/v1/deletebackuprequests?resourceVersion=0: (1m0.00825512s) 504
goroutine 13518 [running]:
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).recordStatus(0xc5584fd0a0, 0x1f8)
	/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:207 +0xdd
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).WriteHeader(0xc5584fd0a0, 0x1f8)
	/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:186 +0x35
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*baseTimeoutWriter).timeout(0xc55853f720, 0xc55834e2d0)
	/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:214 +0xbd
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP(0xc42105de00, 0x81a4e40, 0xc5584fd0a0, 0xc558375200)
	/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:114 +0x28c
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithRequestInfo.func1(0x81a4e40, 0xc5584fd0a0, 0xc558375200)
	/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/requestinfo.go:45 +0x219
net/http.HandlerFunc.ServeHTTP(0xc42285b180, 0x81a4e40, 0xc5584fd0a0, 0xc558375200)
	/usr/local/go/src/net/http/server.go:1942 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/request.WithRequestContext.func1(0x81a4e40, 0xc5584fd0a0, 0xc558375200)
	/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/request/requestcontext.go:110 +0xef
net/http.HandlerFunc.ServeHTTP(0xc42105de20, 0x81a4e40, 0xc5584fd0a0, 0xc558375200)
	/usr/local/go/src/net/http/server.go:1942 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithPanicRecovery.func1(0x81a4e40, 0xc5584fd0a0, 0xc558375200)
	/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/wrap.go:41 +0x11b
net/http.HandlerFunc.ServeHTTP(0xc42105de60, 0x81a5b00, 0xc556895b78, 0xc558375200)
	/usr/local/go/src/net/http/server.go:1942 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.(*APIServerHandler).ServeHTTP(0xc427d149f0, 0x81a5b00, 0xc556895b78, 0xc558375200)
	/workspace/anago-v1.8.10-beta.0.26+044cd262c40234/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/handler.go:198 +0x51
net/http.serverHandler.ServeHTTP(0xc4293c8630, 0x81a5b00, 0xc556895b78, 0xc558375200)
	/usr/local/go/src/net/http/server.go:2568 +0x92
net/http.initNPNRequest.ServeHTTP(0xc42a71bc00, 0xc4293c8630, 0x81a5b00, 0xc556895b78, 0xc558375200)
	/usr/local/go/src/net/http/server.go:3088 +0x93
net/http.(*initNPNRequest).ServeHTTP(0xc4b4d968c0, 0x81a5b00, 0xc556895b78, 0xc558375200)
	<autogenerated>:312 +0x74
net/http.(Handler).ServeHTTP-fm(0x81a5b00, 0xc556895b78, 0xc558375200)
	/usr/local/go/src/net/http/h2_bundle.go:4331 +0x4d
net/http.(*http2serverConn).runHandler(0xc4b4d8c8c0, 0xc556895b78, 0xc558375200, 0xc55853f6c0)
	/usr/local/go/src/net/http/h2_bundle.go:4611 +0x89
created by net/http.(*http2serverConn).processHeaders
	/usr/local/go/src/net/http/h2_bundle.go:4343 +0x4ac

logging error output: "{\"metadata\":{},\"status\":\"Failure\",\"message\":\"Timeout: request did not complete within 1m0s\",\"reason\":\"Timeout\",\"details\":{},\"code\":504}\n"
 [[kube-controller-manager/v1.8.10 (linux/amd64) kubernetes/044cd26/system:serviceaccount:kube-system:generic-garbage-collector] 127.0.0.1:37606]
I0608 10:26:05.057118       7 trace.go:76] Trace[1657744564]: "Get /api/v1/namespaces/kube-system/endpoints/kube-controller-manager" (started: 2018-06-08 10:26:03.549633214 +0000 UTC) (total time: 1.503589434s):
```

Because I haven't have time to trace all this calls and the cluster didn't recover, I removed all the ark keys from etcd when finally the cluster recovered and I my pain was gone ;)

## Next steps?

I still have the backups from ark on s3 so we can go further on this investigation, I noticed that ark never deleted the `deletebackuprequests.ark.heptio.com` CRD, maybe a garbage collector on this CRD can sort it.

```
$ ls -Rl deletebackuprequests.ark.heptio.com/|wc -l
   18761
````

As I said before I'm not sure that ark caused the issue or if it's a k8s crd bug, I'd like to share my experience to avoid users stumbling in the same issue.

## Cluster specs

k8s version **1.8.10**
kops version **1.8.1**
etcd **2.2.1**

**Comment by ncdc** [comment]
@gianrubio thanks for filing this. We'll review it and respond with feedback in a bit. I did want to quickly point out that we do have built-in GC on the `deletebackuprequests` - every hour, the Ark server deletes any of them that are older than 24 hours.

**Comment by ncdc** [comment]
@gianrubio I discussed with some team members, and we think the first thing to address is etcd. Are you able to upgrade to 3.x? If not, we'd recommend you evaluate etcd 2's memory limit and consider increasing it.

**Comment by rosskukulinski** [comment]
@ncdc - I'm a little confused -- [00-ark-config.yaml](https://github.com/heptio/ark/blob/master/examples/aws/00-ark-config.yaml) doesn't have any impact on how often or how many backups/snapshots are taken, right?  That would actually be controlled by defining a backup schedule.  Out of the box, Ark doesn't start taking snapshots by default.

**Comment by ncdc** [comment]
@rosskukulinski you're correct; I hadn't noticed that particular part of the description.

@gianrubio Ark doesn't create any scheduled backups when you install it - you have to do that manually. The default 30 minute `backupSyncPeriod` in the config refers to how often the Ark server should check object storage for backups that might be missing from the Kubernetes API.

Did you create a `schedule` to run every 30 minutes?

**Comment by skriss** [comment]
Closing as there are no further action items here.

**Comment by omegang610** [comment]
@gianrubio 
Do you mind to share what was the root cause of the outage? i am kind of confused 
Thanks
